memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/SessionTokenManager.java

```diff
@@ -0,0 +1,162 @@
+        .build();
+    secureClient = HttpClient.create(provider).option(ChannelOption.SO_SNDBUF, 4 * 1024 * 1024)
+        .option(ChannelOption.SO_LINGER, 0).secure();
+  }
+
+  public static SessionTokenManager getInstance() {
+    return mgr;
+  }
+
+  public SessionCreds getCredentials(final String bucketname) throws InterruptedException {
+    ConcurrentLinkedDeque<SessionCreds> concurrentLinkedDeque = bucketCredentialMap.get(bucketname);
+    if (concurrentLinkedDeque == null) {
+      synchronized (bucketCredentialMap) {
+        concurrentLinkedDeque = bucketCredentialMap.get(bucketname);
+        if (concurrentLinkedDeque == null) {
+          concurrentLinkedDeque = new ConcurrentLinkedDeque<>();
+          // start the scheduled task for credential refresh
+          final ConcurrentLinkedDeque<SessionCreds> concurrentLinkedDequeRef = concurrentLinkedDeque;
+          es.scheduleAtFixedRate(() -> {
+            try {
+              SessionCreds fetchCredentials = fetchCredentials(bucketname);
+              concurrentLinkedDequeRef.add(fetchCredentials);
+              if (concurrentLinkedDequeRef.size() == 2) {
+                // purge existing credentials
+                concurrentLinkedDequeRef.poll();
+              }
+            } catch (Exception e) {
+              e.printStackTrace();
+            }
+          }, 0, 4, TimeUnit.MINUTES);
+          bucketCredentialMap.put(bucketname, concurrentLinkedDeque);
+        }
+      }
+    }
+    while (concurrentLinkedDeque.isEmpty()) {
+      Thread.sleep(100);
+    }
+    return concurrentLinkedDeque.peek();
+  }
+
+  protected SessionCreds fetchCredentials(String bucketname) throws Exception {
+    InstanceProfileCredentialsProvider credentialProvider = InstanceProfileCredentialsProvider
+        .builder().asyncCredentialUpdateEnabled(true).asyncThreadName("IamCredentialUpdater")
+        .build();
+    // Sign it...
+    AwsS3V4Signer signer = AwsS3V4Signer.create();
+    SdkHttpFullRequest req = SdkHttpFullRequest.builder()
+        .appendHeader("x-amz-create-session-mode", "ReadWrite")
+        .appendRawQueryParameter("session", "").method(SdkHttpMethod.GET)
+        .uri(URI.create("https://" + bucketname + ".s3express-use1-az5.us-east-1.amazonaws.com"))
+        .build();
+    SdkHttpFullRequest req1 = signer.sign(req,
+        AwsS3V4SignerParams.builder().awsCredentials(credentialProvider.resolveCredentials())
+            .signingName("s3express").signingRegion(Region.US_EAST_1).build());
+
+    Map<String, List<String>> headers2 = req1.headers();
+    String awsResponse = secureClient.headers(headers -> {
+      for (Entry<String, List<String>> entry : headers2.entrySet()) {
+        headers.set(entry.getKey(), entry.getValue().get(0));
+      }
+    }).get().uri(req.getUri()).responseSingle((response, bytes) -> bytes.asString()).block();
+    
+    System.out.println("CredentialResponse\n"+awsResponse+"\n\n");
```

Comment:
Suggest to add response code check. This call is using InstanceProfileCredentialsProvider, but the instance may miss s3express:CreateSession permission.

Reply:
Good point, will update.